pkg/cloud/azure/azure_kms.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package azure
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	kms "github.com/Azure/azure-sdk-for-go/sdk/keyvault/azkeys"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/errors"
+)
+
+const (
+	azScheme = "azure-kms"
+
+	AzureVaultName = "AZURE_VAULT_NAME"
+)
+
+type azureKMS struct {
+	kms                      *kms.Client
+	customerMasterKeyID      string
+	customerMasterKeyVersion string
+}
+
+var _ cloud.KMS = &azureKMS{}
+var encryptionAlgorithm = kms.JSONWebKeyEncryptionAlgorithmRSAOAEP256
+
+func init() {
+	cloud.RegisterKMSFromURIFactory(MakeAzureKMS, azScheme)
+}
+
+type kmsURIParams struct {
+	vaultName    string
+	environment  string
+	clientID     string
+	clientSecret string
+	tenantID     string
+}
+
+// resolveKMSURIParams parses the `kmsURI` for all the supported KMS parameters.
+func resolveKMSURIParams(kmsURI cloud.ConsumeURL) (kmsURIParams, error) {
+	params := kmsURIParams{
+		vaultName:    kmsURI.ConsumeParam(AzureVaultName),
+		environment:  kmsURI.ConsumeParam(AzureEnvironmentKeyParam),
+		clientID:     kmsURI.ConsumeParam(AzureClientIDParam),
+		clientSecret: kmsURI.ConsumeParam(AzureClientSecretParam),
+		tenantID:     kmsURI.ConsumeParam(AzureTenantIDParam),
+	}
+
+	// Validate that all the passed in parameters are supported.
+	if unknownParams := kmsURI.RemainingQueryParams(); len(unknownParams) > 0 {
+		return kmsURIParams{}, errors.Errorf(
+			`unknown KMS query parameters: %s`, strings.Join(unknownParams, ", "))
+	}
+
+	return params, nil
+}
+
+func MakeAzureKMS(ctx context.Context, uri string, env cloud.KMSEnv) (cloud.KMS, error) {
+	if env.KMSConfig().DisableOutbound {
+		return nil, errors.New("external IO must be enabled to use KMS")
+	}
+	kmsURI, err := url.ParseRequestURI(uri)
+	if err != nil {
+		return nil, err
+	}
+	if kmsURI.Path == "/" {
+		return nil, errors.Newf("path component of the KMS cannot be empty; must contain the Customer Managed Key")
+	}
+
+	kmsConsumeURL := cloud.ConsumeURL{URL: kmsURI}
+	// Extract the URI parameters required to setup the Azure KMS session.
+	kmsURIParams, err := resolveKMSURIParams(kmsConsumeURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if kmsURIParams.vaultName == "" || kmsURIParams.clientID == "" || kmsURIParams.clientSecret == "" || kmsURIParams.tenantID == "" {
```

Comment:
nit: It also may be easier to loop over these params and use `strings.Builder` to buffer up an error message with all the missing params and return that to the client. That way, you can tell the client exactly what params they are missing.

Reply:
👍 I think a slice + strings.Join might work a bit better here for the comma separator but it's much easier this way, thanks! (I actually spend a few minutes staring at a URI trying to figure out what I had forgotten so this is a great idea.)